### PR TITLE
[java] Name disambiguation in annotations

### PR DESCRIPTION
**Pull request by oowekyala**
## Describe the PR

Implement name disambiguation for annotations. This is implemented by adding a ClassOrInterfaceType inside the annotation, so that the logic is reused. This means, ClassOrInterfaceType is used everywhere a type (or type name, not the same thing) is explicitly written (apart from import statements).

This also removes/deprecates some API we don't need anymore:
* Annotation#getAnnotationName: annotations now have a symbol, which is much more useful and powerful than just this name
* ClassOrInterfaceType#isAnonymousClass: this is just meaningless now
* ClassOrInterfaceType#getImage: now always returns null. There is still getSimpleName, or better, the symbol. 
* ASTType#getTypeImage (deprecated for now): with type resolution there will be more useful ways to pretty print the type. The type image is used to implement rules right now, it would be much more robust to use symbols or subtyping checks instead, when type resolution is done.
* ASTThrowStatement#getFirstClassOrInterfaceTypeImage: this is too specific and quite frankly, useless. The only usage was in ExceptionForFlowControl. This rule compares the name of a thrown exception with the name of a caught exception. It doesn't work unless the name is explicitly written (`throw new E()`). When we have full type resolution we can just use the static type of the expression. The rule is rewritten to compare symbols for now.
* ClassOrInterfaceType#isReferenceToClassSameCompilationUnit (deprecated for now): this is used to implement a fallback when type resolution fails (because of an incomplete classpath). With the new type resolution we won't need it, as if the referenced symbol is declared in the same file, we use the AST in place of a Class transparently.

The validation of disambiguated types is stricter now (eg we report an error if you write `List<A, B>`)

I made sure that ExceptionForFlowControl passes tests, but this will only be true when  #2664 is merged

## Ready?

- [x] Added unit tests for fixed bug/feature
- [x] Passing all unit tests
- [x] Complete build `./mvnw clean verify` passes (checked automatically by travis)
- [x] Added (in-code) documentation (if needed)

### CODE REVIEW

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">No java rules are changed!</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTAnnotation.java [hunk 1]
```diff
@@ -9,45 +9,47 @@
...
         super(id);
     }
 
 
     /**
-     * Returns the name of the annotation as it is used,
-     * eg {@code java.lang.Override} or {@code Override}.
+     * Returns the node that represents the name of the annotation.
      */
-    public String getAnnotationName() {
```
Comment (adangel) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTAnnotation.java:36]: Would it make sense to keep a deprecated `getAnnotationName()` method around (DeprecatedUntil700)?

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTThrowStatement.java [hunk 2]
```diff
@@ -32,23 +32,4 @@ public ASTExpression getExpr() {
...
-     * throw this.e or throw this.
-     *
-     * This is too specific
-     *
-     * <p>TODO - use symbol table (?)</p>
-     *
-     * @return the image of the first ASTClassOrInterfaceType node found or
-     *     <code>null</code>
-     */
-    public String getFirstClassOrInterfaceTypeImage() {
```
Comment (adangel) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTThrowStatement.java:50]: I guess, we should deprecate this method on master. It seems to be used only by one rule - ExceptionAsFlowControl

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/internal/AnnotationSuppressionUtil.java [hunk 3]
```diff
@@ -107,9 +106,8 @@ private static boolean hasSuppressWarningsAnnotationFor(final Annotatable node,
 
     // @formatter:on
     private static boolean annotationSuppresses(ASTAnnotation annotation, Rule rule) {
-        // if (SuppressWarnings.class.equals(getType())) { // typeres is not always on
-        if (TypeHelper.isA(annotation, SuppressWarnings.class)) {
-            for (ASTLiteral element : annotation.findDescendantsOfType(ASTLiteral.class)) {
+        if (annotation.getSymbol().getBinaryName().equals("java.lang.SuppressWarnings")) {
```
Comment (adangel) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/internal/AnnotationSuppressionUtil.java:109]: We'll get a PositionLiteralFirstInComparison violation

```suggestion
        if ("java.lang.SuppressWarnings".equals(annotation.getSymbol().getBinaryName())) {
```

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTThrowStatement.java (hunk 2 above)
Comment (adangel) [reply pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/ASTThrowStatement.java:50]: Done: https://github.com/pmd/pmd/commit/a5eb60243dfa12d3ca2cc8eb3bfff091654c66ab

File: pmd-java/src/test/kotlin/net/sourceforge/pmd/lang/java/ast/TypeDisambiguationTest.kt [hunk 4]
```diff
@@ -133,4 +137,101 @@ class TypeDisambiguationTest : ParserTestSpec({
...
+                acu.descendants(ASTAnnotation::class.java).map { it.typeNode }.toList()
+
+        fun assertErrored(t: ASTClassOrInterfaceType) {
+            val errs = logger.errors[SemanticChecksLogger.EXPECTED_ANNOTATION_TYPE]?.filter { it.first == t } ?: emptyList()
+            assertEquals(errs.size, 1, "`${t.text}` should have produced a single error")
+            errs.single().second.toList() shouldBe emptyList()
+        }
+
+        fun assertNoError(t: ASTClassOrInterfaceType) {
+            val err = logger.errors[SemanticChecksLogger.MALFORMED_GENERIC_TYPE]?.firstOrNull { it.first == t }
```
Comment (oowekyala) [inline pmd-java/src/test/kotlin/net/sourceforge/pmd/lang/java/ast/TypeDisambiguationTest.kt:226]:
```suggestion
            val err = logger.errors[SemanticChecksLogger.EXPECTED_ANNOTATION_TYPE]?.firstOrNull { it.first == t }
```

**Pull request merged**